File: src/main/java/org/littletonrobotics/frc2025/subsystems/superstructure/SuperstructureState.java
// Copyright (c) 2025 FRC 6328
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file at
// the root directory of this project.

package org.littletonrobotics.frc2025.subsystems.superstructure;

import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.littletonrobotics.frc2025.subsystems.superstructure.SuperstructurePose.Preset;
import org.littletonrobotics.frc2025.subsystems.superstructure.SuperstructureStateData.Height;
import org.littletonrobotics.frc2025.subsystems.superstructure.dispenser.Dispenser;

@Getter
@RequiredArgsConstructor
public enum SuperstructureState {
  START(SuperstructureStateData.builder().build()),
  AUTO_START(SuperstructureStateData.builder().build()),
  CHARACTERIZATION(SuperstructureStateData.builder().build()),
  SAFETY(SuperstructureStateData.builder().build()),
  STOW(SuperstructureStateData.builder().pose(Preset.STOW.getPose()).build()),
  CORAL_INTAKE(
      SuperstructureStateData.builder()
          .pose(Preset.CORAL_INTAKE.getPose())
          .tunnelVolts(Dispenser.tunnelIntakeVolts)
          .build()),
  GOODBYE_CORAL(SuperstructureStateData.builder().pose(Preset.GOODBYE_CORAL.getPose()).build()),
  L1_CORAL(SuperstructureStateData.builder().pose(Preset.L1.getPose()).build()),
  L2_CORAL(
      SuperstructureStateData.builder()
          .pose(Preset.L2.getPose())
          .height(Height.FIRST_STAGE)
          .build()),
  L3_CORAL(
      SuperstructureStateData.builder()
          .pose(Preset.L3.getPose())
          .height(Height.FIRST_STAGE)
          .build()),
  L4_CORAL(
      SuperstructureStateData.builder()
          .pose(Preset.L4.getPose())
          .height(Height.SECOND_STAGE)
          .build()),
  GOODBYE_CORAL_EJECT(
      GOODBYE_CORAL.getValue().toBuilder().tunnelVolts(Dispenser.tunnelDispenseVolts[3]).build()),
  L1_CORAL_EJECT(
      SuperstructureStateData.builder()
          .pose(Preset.L1_EJECT.getPose())
          .tunnelVolts(Dispenser.tunnelDispenseVolts[0])
          .build()),
  L2_CORAL_EJECT(
      L2_CORAL.getValue().toBuilder().tunnelVolts(Dispenser.tunnelDispenseVolts[1]).build()),
  L3_CORAL_EJECT(
      L3_CORAL.getValue().toBuilder().tunnelVolts(Dispenser.tunnelDispenseVolts[2]).build()),
  L4_CORAL_EJECT(
      L4_CORAL.getValue().toBuilder().tunnelVolts(Dispenser.tunnelDispenseVolts[3]).build()),
  ALGAE_STOW(
      SuperstructureStateData.builder()
          .pose(Preset.ALGAE_STOW.getPose())
          .gripperGoal(Dispenser.GripperGoal.GRIP)
          .build()),
  ALGAE_CORAL_INTAKE(
      CORAL_INTAKE.getValue().toBuilder().gripperGoal(Dispenser.GripperGoal.GRIP).build()),
  ALGAE_GOODBYE_CORAL(
      GOODBYE_CORAL.getValue().toBuilder().gripperGoal(Dispenser.GripperGoal.GRIP).build()),
  ALGAE_L2_CORAL(
      SuperstructureStateData.builder()
          .pose(Preset.ALGAE_L2.getPose())
          .gripperGoal(Dispenser.GripperGoal.GRIP)
          .height(Height.FIRST_STAGE)
          .build()),
  ALGAE_L3_CORAL(
      SuperstructureStateData.builder()
          .pose(Preset.ALGAE_L3.getPose())
          .gripperGoal(Dispenser.GripperGoal.GRIP)
          .height(Height.FIRST_STAGE)
          .build()),
  ALGAE_L4_CORAL(
      SuperstructureStateData.builder()
          .pose(Preset.ALGAE_L4.getPose())
          .gripperGoal(Dispenser.GripperGoal.GRIP)
          .height(Height.SECOND_STAGE)
          .build()),
  ALGAE_GOODBYE_CORAL_EJECT(
      ALGAE_GOODBYE_CORAL.getValue().toBuilder()
          .tunnelVolts(Dispenser.tunnelDispenseVolts[3])
          .build()),
  ALGAE_L2_CORAL_EJECT(
      ALGAE_L2_CORAL.getValue().toBuilder()
          .tunnelVolts(Dispenser.tunnelDispenseVoltsAlgae[0])
          .build()),
  ALGAE_L3_CORAL_EJECT(
      ALGAE_L3_CORAL.getValue().toBuilder()
          .tunnelVolts(Dispenser.tunnelDispenseVoltsAlgae[1])
          .build()),
  ALGAE_L4_CORAL_EJECT(
      ALGAE_L4_CORAL.getValue().toBuilder()
          .tunnelVolts(Dispenser.tunnelDispenseVoltsAlgae[2])
          .build()),
  ALGAE_L2_INTAKE(
      SuperstructureStateData.builder()
          .pose(Preset.ALGAE_L2_INTAKE.getPose())
          .gripperGoal(Dispenser.GripperGoal.GRIP)
          .height(Height.FIRST_STAGE)
          .build()),
  ALGAE_L3_INTAKE(
      SuperstructureStateData.builder()
          .pose(Preset.ALGAE_L3_INTAKE.getPose())
          .gripperGoal(Dispenser.GripperGoal.GRIP)
          .height(Height.FIRST_STAGE)
          .build()),
  ALGAE_ICE_CREAM_INTAKE(
      SuperstructureStateData.builder()
          .pose(Preset.ALGAE_ICE_CREAM_INTAKE.getPose())
          .gripperGoal(Dispenser.GripperGoal.GRIP)
          .build()),
  PRE_THROW(
      SuperstructureStateData.builder()
          .pose(Preset.THROW.getPose())
          .gripperGoal(Dispenser.GripperGoal.GRIP)
          .height(Height.SECOND_STAGE)
          .build()),
  THROW(PRE_THROW.getValue().toBuilder().gripperGoal(Dispenser.GripperGoal.EJECT).build()),
  TOSS(
      SuperstructureState.ALGAE_STOW.getValue().toBuilder()
          .gripperGoal(Dispenser.GripperGoal.EJECT)
          .build()),
  ALGAE_STOW_INTAKE(SuperstructureState.ALGAE_STOW.getValue().toBuilder().build()),
  PRE_PROCESS(
      SuperstructureStateData.builder()
          .pose(Preset.PROCESS.getPose())
          .gripperGoal(Dispenser.GripperGoal.GRIP)
          .build()),
  PROCESS(
      SuperstructureState.PRE_PROCESS.getValue().toBuilder()
          .gripperGoal(Dispenser.GripperGoal.EJECT)
          .build());

  private final SuperstructureStateData value;
}

File: src/main/cpp/subsystems/superstructure/SuperstructureState.cpp
// Copyright (c) 2025 FRC 6328
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file at
// the root directory of this project.

#include "subsystems/superstructure/SuperstructureState.h"

#include <optional>

#include <frc/geometry/Rotation2d.h>

#include "subsystems/superstructure/SuperstructurePose.h"
#include "subsystems/superstructure/SuperstructureStateData.h"
#include "subsystems/superstructure/dispenser/Dispenser.h"

SuperstructureStateData GetValue(SuperstructureState state) {
  switch (state) {
  case SuperstructureState::START:
    return SuperstructureStateData::BuilderCreate().Build();
  case SuperstructureState::AUTO_START:
    return SuperstructureStateData::BuilderCreate().Build();
  case SuperstructureState::CHARACTERIZATION:
    return SuperstructureStateData::BuilderCreate().Build();
  case SuperstructureState::STOW:
    return SuperstructureStateData::BuilderCreate()
        .Pose(SuperstructurePose(SuperstructurePose::Preset::STOW))
        .Build();
  case SuperstructureState::INTAKE:
    return SuperstructureStateData::BuilderCreate()
        .Pose(SuperstructurePose(SuperstructurePose::Preset::INTAKE))
        .TunnelVolts(Dispenser::tunnelIntakeVolts)
        .Build();
  case SuperstructureState::GOODBYE_CORAL:
    return SuperstructureStateData::BuilderCreate()
        .Pose(SuperstructurePose(SuperstructurePose::Preset::GOODBYE_CORAL))
        .Build();
  case SuperstructureState::L1_CORAL:
    return SuperstructureStateData::BuilderCreate()
        .Pose(SuperstructurePose(SuperstructurePose::Preset::L1))
        .Build();
  case SuperstructureState::L2_CORAL:
    return SuperstructureStateData::BuilderCreate()
        .Pose(SuperstructurePose(SuperstructurePose::Preset::L2))
        .Height(SuperstructureStateData::Height::FIRST_STAGE)
        .Build();
  case SuperstructureState::L3_CORAL:
    return SuperstructureStateData::BuilderCreate()
        .Pose(SuperstructurePose(SuperstructurePose::Preset::L3))
        .Height(SuperstructureStateData::Height::FIRST_STAGE)
        .Build();
  case SuperstructureState::L4_CORAL:
    return SuperstructureStateData::BuilderCreate()
        .Pose(SuperstructurePose(SuperstructurePose::Preset::L4))
        .Height(SuperstructureStateData::Height::SECOND_STAGE)
        .Build();
  case SuperstructureState::GOODBYE_CORAL_EJECT:
    return GetValue(SuperstructureState::GOODBYE_CORAL)
        .ToBuilder()
        .TunnelVolts(Dispenser::tunnelDispenseVolts)
        .Build();
  case SuperstructureState::L1_CORAL_EJECT:
    return SuperstructureStateData::BuilderCreate()
        .Pose(SuperstructurePose(SuperstructurePose::Preset::L1_EJECT))
        .TunnelVolts(Dispenser::tunnelL1DispenseVolts)
        .GripperGoal(Dispenser::GripperGoal::L1_EJECT)
        .Build();
  case SuperstructureState::L2_CORAL_EJECT:
    return GetValue(SuperstructureState::L2_CORAL)
        .ToBuilder()
        .TunnelVolts(Dispenser::tunnelDispenseVolts)
        .GripperGoal(Dispenser::GripperGoal::L1_EJECT)
        .Build();
  case SuperstructureState::L3_CORAL_EJECT:
    return GetValue(SuperstructureState::L3_CORAL)
        .ToBuilder()
        .TunnelVolts(Dispenser::tunnelDispenseVolts)
        .Build();
  case SuperstructureState::L4_CORAL_EJECT:
    return GetValue(SuperstructureState::L4_CORAL)
        .ToBuilder()
        .TunnelVolts(Dispenser::tunnelDispenseVolts)
        .GripperGoal(Dispenser::GripperGoal::EJECT)
        .Build();
  case SuperstructureState::ALGAE_L2_INTAKE:
    return SuperstructureStateData::BuilderCreate()
        .Pose(SuperstructurePose(SuperstructurePose::Preset::ALGAE_L2_INTAKE))
        .GripperGoal(Dispenser::GripperGoal::GRIP)
        .Height(SuperstructureStateData::Height::FIRST_STAGE)
        .Build();
  case SuperstructureState::ALGAE_L3_INTAKE:
    return SuperstructureStateData::BuilderCreate()
        .Pose(SuperstructurePose(SuperstructurePose::Preset::ALGAE_L3_INTAKE))
        .GripperGoal(Dispenser::GripperGoal::GRIP)
        .Height(SuperstructureStateData::Height::FIRST_STAGE)
        .Build();
  case SuperstructureState::PRE_THROWN:
    return SuperstructureStateData::BuilderCreate()
        .Pose(SuperstructurePose(SuperstructurePose::Preset::PRE_THROW))
        .GripperGoal(Dispenser::GripperGoal::GRIP)
        .Height(SuperstructureStateData::Height::SECOND_STAGE)
        .Build();
  case SuperstructureState::THROWN:
    return SuperstructureStateData::BuilderCreate()
        .Pose(SuperstructurePose(SuperstructurePose::Preset::THROW))
        .GripperGoal(Dispenser::GripperGoal::EJECT)
        .Build();
  case SuperstructureState::ALGAE_STOW:
    return SuperstructureStateData::BuilderCreate()
        .Pose(SuperstructurePose(SuperstructurePose::Preset::ALGAE_STOW))
        .GripperGoal(Dispenser::GripperGoal::GRIP)
        .Build();
  case SuperstructureState::PRE_TOSS:
    return SuperstructureStateData::BuilderCreate()
        .Pose(SuperstructurePose(SuperstructurePose::Preset::ALGAE_STOW)
                  .elevatorHeight,
              []() { return frc::Rotation2d::Degrees(30.0); })
        .GripperGoal(Dispenser::GripperGoal::GRIP)
        .Build();
  case SuperstructureState::TOSS:
    return GetValue(SuperstructureState::PRE_TOSS)
        .ToBuilder()
        .GripperGoal(Dispenser::GripperGoal::EJECT)
        .Build();
  case SuperstructureState::ALGAE_STOW_INTAKE:
    return GetValue(SuperstructureState::ALGAE_STOW).ToBuilder().Build();
  case SuperstructureState::PROCESSED:
    return GetValue(SuperstructureState::ALGAE_STOW)
        .ToBuilder()
        .GripperGoal(Dispenser::GripperGoal::EJECT)
        .Build();
  default:
    return SuperstructureStateData::BuilderCreate().Build();
  }
}